Use correct MIME type for multipart form data.

eehakkin opened the pull request:
Objects implementing the Body mixin return a new FormData object only if
a MIME type is a correct one and throw a TypeError otherwise. Thus use
a correct MIME type for formData tests.

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -21,7 +29,7 @@
     function checkBodyBlob(response, expectedBody, checkContentType) {
       return response.blob().then(function(bodyAsBlob) {
-        if (checkContentType)
+        if (checkContentType !== false)
```
This change is ok, but why should it be done? Is it the expected wpt coding style?

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -58,61 +66,68 @@
 
     function checkBodyFormData(response, expectedBody) {
       return response.formData().then(function(bodyAsFormData) {
-        assert_true(bodyAsFormData instanceof FormData, "Should receive a FormData");
```
This assert is still true, why removing it?

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -58,61 +66,68 @@
-    formData.append("name", "value");
     var textData = JSON.stringify("This is response's body");
-    var blob = new Blob([textData], { "type" : "text/plain" });
+    var blob = new Blob([textData], { "type": "application/octet-stream" });
```
This is ok but is there a particular reason to change the type here?

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -58,61 +66,68 @@
+    var blob = new Blob([textData], { "type": "application/octet-stream" });
+    var blobResponseInit = { "headers": [["Content-Type", "text/PLAIN"]] };
+    var multipartBoundary = "boundary-" + Math.random();
+    var formResponseInit = { "headers": [["Content-Type", 'multipart/FORM-data; boundary="' + multipartBoundary + '"']] };
```
Would it make sense to add a test with a bad header value?

eehakkin replied:
Yes, it would make sense. Done.

youennf reviewed (commented):
Sounds good to me.
A couple of questions and it would be great to add an error test for bad header values

eehakkin committed:
Use correct MIME type for multipart form data.

Objects implementing the Body mixin return a new FormData object only if
a MIME type is a correct one and throw a TypeError otherwise. Thus use
a correct MIME type for formData tests. Additionally, improve existing
formData tests and add new ones.

eehakkin commented:
I rebased on top of master, resolved conflicts caused by #3948 and improved based on that commit. Please take a look.

eehakkin commented:
@youennf, can I do something to get this merged?

youennf commented:
I forgot about this one... sorry about that.
I will check it today.

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -12,17 +12,31 @@
+
+    function stringToMultipartFormTextData(multipartBoundary, name, value) {
+      return ('--' + multipartBoundary + '\r\n' +
+              'Content-Disposition: form-data; name="' + name + '"\r\n' +
```
I guess this is ok for now.
I wonder whether we will not want something like:
Content-Disposition: form-data;name...

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -56,80 +70,134 @@
 
-    function checkResponseBody(body, bodyType, bodyConsumingType, checkFunction) {
+    function checkBodyFormDataError(response, expectedBody) {
+      return response.formData().then(
```
Can you use promise_rejects instead?

eehakkin replied:
Done. I was trying to avoid passing the test object to check function.

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -56,80 +70,134 @@
+        });
+    }
+
+    function checkResponseBody(body, expectedBody, bodyType, bodyConsumingType, checkFunction, responseInit) {
```
That makes a lot of parameters, making it a bit difficult to understand.
How about moving bodyType and bodyConsumeType at the end.
And maybe combining them in something like "from XX to YY", the response name being 
"Consume response body: from XX to YY"

Also we use have two parameters body and responseInit. Might want to group them together and put them at the start. Or directly pass a response if that improves readability.

eehakkin replied:
> How about moving bodyType and bodyConsumeType at the end.
> And maybe combining them in something like "from XX to YY", the response name being
> "Consume response body: from XX to YY"

Done.

> Also we use have two parameters body and responseInit. Might want to group them together and put them at the start. Or directly pass a response if that improves readability.

I'd rather not pass a response directly because that would change initialization errors from test failures to global errors. However, I created helper functions which create response promises so that the two parameters are grouped together and named logically.

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -56,80 +70,134 @@
+    checkResponseBody(urlSearchParamsBlob, urlSearchParamsData, "blob", "formData with correct urlencoded type", checkBodyFormDataUrlencoded, urlSearchParamsResponseInit);
+    checkResponseBody(urlSearchParamsBlob, undefined, "blob", "formData without correct urlencoded type", checkBodyFormDataError, textResponseInit);
+
+    function checkFormDataResponseBody(body, expectedName, expectedValue, bodyType, bodyConsumingType, checkFunction, responseInit) {
```
Ditto here

youennf commented on fetch/api/response/response-consume.html:
```diff
@@ -56,80 +70,134 @@
     checkFetchedResponseBody("blob", checkBodyBlob);
     checkFetchedResponseBody("text", checkBodyText);
     checkFetchedResponseBody("arrayBuffer", checkBodyArrayBuffer);
+    checkFetchedResponseBody("formData without correct type", checkBodyFormDataError);
```
It might not be very clear for readers that some tests are error cases (meaning we will not be able to get a body) and some are regular cases. Maybe we should group all the error tests together or give them a name suffix like (error case).

eehakkin replied:
I added the name suffix.

youennf commented:
Sounds good in general.
I am just wondering how we could improve a bit the readability. See comments above.

eehakkin commented:
With combined parameters and named helper functions, I think that the readability has improved.

youennf commented:
Sounds great, thanks!
